### PR TITLE
feat: Preserve symlinks when copying templates

### LINKED ISSUE #937: allow to choose if symlinks should be copied as symlinks
**Describe the problem**
I have a copier template that has a `CHANGELOG.rst` file in the template root and a symlink to it in its `docs/` folder. After running `copier copy` the generated project ends up with a second copy of `CHANGELOG.rst` in the `docs/` folder instead of a symlink to it.

**Template**

https://github.com/AdrianFreundQC/copier-symlink-demo
Also as a zip here: [copier-symlink-demo-main.zip](https://github.com/copier-org/copier/files/10511794/copier-symlink-demo-main.zip)

**To Reproduce**

1. Run `copier copy https://github.com/AdrianFreundQC/copier-symlink-demo out`
2. Run `ls -l out/`
3. See that `symlink.txt` is not a symlink

**Logs**

<details>

```
$ copier copy https://github.com/AdrianFreundQC/copier-symlink-demo out
No git tags found in template; using HEAD as ref

Copying from template version 0.0.0.post1.dev0+9ac1db8
    create  .
    create  symlink.txt
    create  original_file.txt

$ ls -l out/
total 16
-rw-r--r--@ 1 adrian  staff  12 Jan 26 17:33 original_file.txt
-rw-r--r--@ 1 adrian  staff  12 Jan 26 17:33 symlink.txt
```

</details>

**Expected behavior**

Copier should preserve the symlink:
```
$ ls -l out/
total 8
-rw-r--r--@ 1 adrian  staff  12 Jan 26 17:33 original_file.txt
lrwxr-xr-x@ 1 adrian  staff  17 Jan 26 17:36 symlink.txt -> original_file.txt
```

**Environment**
 - OS: macOS Ventura 13.1
 - Copier version: copier 7.0.1
 - Python version: Python 3.11.0
 - Installation method: mamba+conda-forge

### PR DESCRIPTION
**Pull request by AdrianFreundQC**
Closes #937.

This PR makes copier preserve symlinks when copying a template instead of copying the file the symlink points to.

The old behavior can be restored by setting `_preserve_symlinks` to false in `copier.yml`.

If the name of a symlink is templated it will be rendered. The targets of symlinks will only be rendered if the symlink name ends with the template suffix (`.jinja` by default).

### CODE REVIEW

**Comment by AdrianFreundQC** [comment]
I have just force-pushed some fixed. I can now confirm that this works using the template in the linked issue and breaks no existing tests.

I haven't yet written automated tests for this.

**Comment by yajo** [comment]
I think we'll break a lot of workflows if this is not optional (even if it's enabled by default, I think there should be a way to disable it).

@sisp @pawamoy what do you think?

**Comment by pawamoy** [comment]
I don't know about a lot of workflows, but just recently we had someone with a use-case for symlinks being rendered.

**Comment by AdrianFreundQC** [comment]
Pushed again. Tests should no longer fail on python <3.10.

I can try making this an option. Probably only next week though.

**Comment by sisp** [comment]
> This is actually a feature that allows you to reuse files in some ways and reduce code duplication. However indeed it would make sense to make it optional.
>
> &mdash; https://github.com/copier-org/copier/issues/937#issuecomment-1405775180

TBH, I would have expected that symlinks remain symlinks after generating a project. Would it be possible to "reuse files [...] and reduce code duplication" via [Jinja "include"](https://jinja.palletsprojects.com/en/3.1.x/templates/#include) instead? If it works, it only works for text files though but not for binary files. Or [Copier "tasks"](https://copier.readthedocs.io/en/latest/configuring/#tasks) could be used to copy files, although I admit it's less declarative than the current symlink approach or the Jinja "include" idea.

If making it optional is the way to go, I wonder whether it's sufficient to have a global setting or whether it should be configurable at a more granular level. :thinking:

**Comment by sisp** [comment]
By the way, the current symlink behavior seems to be undocumented. This leads me to wonder:

1. How many Copier users actually rely on the current symlink behavior?
2. How bad would it be to introduce a breaking change to this undocumented feature? (Although it's certainly not a good idea to break people's templates, but relying on an undocumented feature is known to be dangerous.)

**Comment by AdrianFreundQC** [comment]
My suggestion would be the following:
Preserve symlinks by default. This is what I would expect and what is in line with other tools such as cookiecutter. If this is too disruptive it might be possible to have it not preserve by default and change the default once 8.0 releases.

jinja templating in both the symlink name and the symlink target should be rendered (So you can for example have a symlink that points to a file with a templated name).

There should be an option in `copier.yml` to change whether or not to preserve symlinks. When it is off everything should behave as before.
I'm not sure if a CLI option should exist, but I tend towards no, as it would probably mostly lead to users breaking templates that were designed with one of the options in mind.

**Comment by AdrianFreundQC** [comment]
I also just checked the top 20 public copier templates on github and found exactly one that relies on symlinks being replaced with the file they point to.

https://github.com/OCA/oca-addons-repo-template uses a symlink to copy the `.gitignore` from the template repo to  generated project.

Maybe a third option would be useful for some (Preserve all symlinks, preserve no symlinks, preserve only symlinks pointing to inside the generated project). Or maybe that would just be confusing and lead to more problems...

**Comment by yajo** [comment]
> https://github.com/OCA/oca-addons-repo-template uses a symlink to copy the `.gitignore` from the template repo to generated project.

I happen to maintain that template, and some others with at least that same purpose. However that can be easily changed to using import.

I'd just add an option to copier.yaml: `_symlinks`  with choices `as_symlinks` or `as_files`.



> jinja templating in both the symlink name and the symlink target should be rendered (So you can for example have a symlink that points to a file with a templated name).

Makes total sense.

File: copier/main.py
Comment (yajo) [inline copier/main.py:534]: imho the symlink should be copied raw always, when `preserve_symlinks == True`.

If it is absolute, it should be copied absolute. If it is relative, then relative. Template maintainers can create the symlink as they want, so it's better to lend them that responsability. Just document it and test it.

File: copier/template.py
Comment (yajo) [inline copier/template.py:445]: > I'd just add an option to copier.yaml: `_symlinks` with choices `as_symlinks` or `as_files`.

Sorry, I didn't see this code. Your option name seems better to me. And I like it more that it is a boolean. Please leave it as you've done, thanks!

File: copier/tools.py
Comment (yajo) [inline copier/tools.py:187]:
```suggestion
    """A custom version of os.readline/pathlib.Path.readlink.

    pathlib.Path.readlink is what we ideally would want to use, but it is only available on python>=3.9.
    os.readlink doesn't support Path and bytes on Windows for python<3.8
```

**Review by yajo — CHANGES_REQUESTED** [review]

**Comment by AdrianFreundQC** [comment]
I think this PR should be ready now.

File: copier/main.py
Comment (yajo) [inline copier/main.py:325]: Seems less error-prone:

```suggestion
                previous_content = readlink(dst_abspath)
```

File: copier/main.py
Comment (yajo) [inline copier/main.py:304]:
```suggestion
        expected_contents: Union[bytes, str] = b"",
```

File: copier/main.py
Comment (yajo) [inline copier/main.py:536]:
```suggestion
            expected_contents=dst_target,
```

File: copier/main.py
Comment (yajo) [inline copier/main.py:522]: I think I commented it previously... Absolute links should be kept absolute IMHO.

File: copier/main.py
Comment (yajo) [inline copier/main.py:520]: Please add a docstring to the method, explaining clearly what it does. It will help my future self when maintaining this 😄

File: copier/main.py
Comment (pawamoy) [reply copier/main.py:530]: Yep, docs absolutely needed, especially since there's a new config option `_preserve_symlinks`.

File: copier/main.py
Comment (AdrianFreundQC) [reply copier/main.py:522]: This doesn't change the target of the symlink. This is the path where the symlink itself is.
The handling of the symlink target is in the code block below

File: copier/main.py
Comment (AdrianFreundQC) [reply copier/main.py:530]: Is it enough to add it to the `preserve_symlinks` option documentation or should I also add it somewhere else in the docs? If so, where would it fit best?

File: copier/main.py
Comment (AdrianFreundQC) [reply copier/main.py]: I used `Union[bytes, Path]` instead.

File: copier/template.py
Comment (yajo) [inline copier/template.py:444]: To fit with style:

```suggestion
        """Know if Copier should preserve symlinks when rendering the template.

        See [preserve_symlinks][].
    """
```

File: copier/template.py
Comment (yajo) [inline copier/template.py:445]: I think that we should default to `False`.

There are a lot of templates that use this feature as it was implemented before. At least I have a few myself. I also think it's a good default.

```suggestion
        return bool(self.config_data.get("preserve_symlinks", False))
```

This will affect some of the tests.

File: docs/configuring.md
Comment (yajo) [inline docs/configuring.md:861]: As above:

```suggestion
-   Default value: `False`
```

File: copier/main.py
Comment (yajo) [reply copier/main.py:540]: I see it OK with the docs you already added. Thanks!

**Review by yajo — CHANGES_REQUESTED** [review]
Thanks for this enhancement. It is very well structured! And special thanks for the tests!

Some comments below.

File: copier/template.py
Comment (AdrianFreundQC) [reply copier/template.py]: Changed default to false and adjusted tests. Most tests test the default (False) behavior. All tests testing the True behavior are in `test_symlinks.py`

**Comment by AdrianFreundQC** [comment]
I addressed all the review comments and rebased because of a conflict. I'm not yet sure why one test fails when run with `flake check`, but passes when run  with `poe test`, but I can reproduce it locally and am looking into it.

**Comment by AdrianFreundQC** [comment]
Tests pass again. nix flake didn't preserve the symlink in the demo template while copying it to the nix environment. The test couldn't preserve the symlink because it wasn't there in the first place.

I removed the symlink from the demo template again and instead added more fine-grained tests, so this has 100% diff coverage again.

I'm not familiar enough with nix to know if this is a bug or expected behavior.

**Comment by AdrianFreundQC** [comment]
I just rebased this to fix the merge conflicts. Any update on when this might get merged or get another review round?

**Review by yajo — APPROVED** [review]
Excellent, thanks!

**Pull request merged**
